### DISCUSSION TITLE
Mission Ported SW

### DISCUSSION THREAD

**Discussion by satzguyz** [nasa/fprime#1365] [Q&A]
Hello,
Is there any reference codebase where the mission specific features are fully loaded. 

Example: Telemetry and Telecommands list, Mission specific applications, etc

**Comment by LeStarch** [comment]
If you are looking for a full example deployment, we are working on that.  It is in the near-term plan.

**Reply by ThibFrgsGmz** [reply]
@LeStarch Interesting, what would that full example consist of? 
Is it a full flight demonstrator that works in space?

**Reply by LeStarch** [reply]
First, this is still in early discussions so it may change. However, the basic idea is to create a system deployment example. It would walk through some of the standard parts of most deployments.  Those are:

1. Telecommunication Adaptation.  A component that communicates over a commercially available radio.  Demo how to integrate a radio into F´.
2. Sensor or Science package.  A component that reads and telemeters data from a sensor or science component. Demo how to work with sensors and hardware in an F´ context.

This is minimal, but most systems are (at their core) a telecom setup that likely isn't UART nor Ip and N sensors.

If you have other thoughts, we'd consider them!

**Reply by LeStarch** [reply]
One of the primary goals is to make the necessary hardware easily accessible. Thus we are likely to stick with a RaspberryPI, a commercial sensor, and a commercial radio.  It likely wouldn't work in space (or not for long).  However, this means many groups can actually build and test with this system as a learning exercise.

**Reply by ThibFrgsGmz** [reply]
Indeed working directly on commercial off-the-shelf hardware can be a valuable learning experience for aerospace and STEM education.

By the way, I wasn't necessarily expecting a software system that required a space-qualified hardware platform. I am purely interested in the software architecture, regardless of the hardware. :)

In that sense, you could use this deployment to take the exercise of designing a flight system even further. For example, implement key components such as FDIR, operating mode management, and memory scrubbing (EDAC).

**Comment by ThibFrgsGmz** [comment]
@LeStarch 

> If you have other thoughts, we'd consider them!

Personally, I'd like to see examples of deployments to design vision-based satellite equipment such as rendezvous cameras, debris detection systems, star trackers, and so on.

I could be wrong, but it seems to me that creating a startup to develop this type of equipment is less difficult than developing a satellite onboard computer.

With the increase in the number of startups offering vision and AI based space services and applications, FPrime has a role to play in providing an environment in which these players can focus on their added value.

As a result, I believe we should be able to create a complete digital twin of the system to increase system reliability, provide more flexible hardware options, reduce time to market for users and accelerate market growth.

To do this, we should move closer to IoT technologies such as [Antmicro](https://antmicro.com/)'s [Renode](https://renode.io/) and open hardware (RISC-V).


Renode is currently of interest to me because it offers integrated software emulation for a wide range of Arm and RISC-V based hardware platforms, as well as simulation of complete sensor nodes.

Renode allows for the creation of different scenarios and the study of different environmental configurations to improve test coverage and, consequently, the quality and reliability of the software system.

Of particular relevance is Renode's emulation of the Microchip/Microsemi [PolarFire Soc](https://www.microsemi.com/product-directory/soc-fpgas/5498-polarfire-soc-fpga), which is based on a Risc-V (Mi-V) CPU.
This component would be a good candidate for a space platform because of its radiation robustness.


In short, I would propose a digital twin of a satellite flight system. It would run on Renode's Polarfire Soc emulation and interacts with a simulated sensor (a camera) as well as a client (the equivalent of the satellite's onboard computer).
The data from the camera would be images, which the software would process to detect contours or apply other processing before sending the data to the on-board computer.


Note: The [42 simulator](https://github.com/ericstoneking/42) (spacecraft attitude and orbit dynamics simulator) used in [OpenSatKit](https://github.com/OpenSatKit/OpenSatKit) might be interesting to implement as well.

**Reply by astroesteban** [reply]
Have there been any efforts in integrating F Prime with the 42 simulator or [NASA's Trick](https://github.com/nasa/trick)? That would be a very interesting and useful development.

**Reply by Joshua-Anderson** [reply]
One issue is that F' doesn't offer a standard GNC sensor interface (IMU, sun sensor, star tracker, GPS, etc) at the moment, so there's no great way to generically interface with a simulation right now. A demo project with mock sensors interfaced with a simulator would be very cool though.

**Reply by astroesteban** [reply]
Yup that would be great. Even just like a simple lookup table for some dummy sensors would be nice. Mostly to get an idea on how to go about interfacing F' with these simulation environments.
